### PR TITLE
adding sentry logging to 500s we are seeing

### PR DESCRIPTION
**Pull request by tejans24**
connects department-of-veterans-affairs/appeals-deployment#1038

### CODE REVIEW

File: app/controllers/annotation_controller.rb [hunk 1]
```diff
@@ -2,6 +2,9 @@ class AnnotationController < ApplicationController
   before_action :verify_access
 
   rescue_from ActiveRecord::RecordInvalid do |e|
+    Rails.logger.error "#{e.message}\n#{e.backtrace.join("\n")}"
+    Raven.capture_exception(e)
```
Comment (nickheiner-usds) [inline app/controllers/annotation_controller.rb]: Why is this necessary? Don't we already catch all exceptions in Rails and pass them to Raven?

Comment (tejans24) [reply app/controllers/annotation_controller.rb]: I think that only works for uncaught exceptions. These are being captured and handled by returning a json response.

Comment (nickheiner-usds) [reply app/controllers/annotation_controller.rb]: That makes sense. Is there anywhere else in the codebase that we have this same problem, then?

Comment (tejans24) [reply app/controllers/annotation_controller.rb]: Yea, I am creating an issue to evaluate other cases where we need to log errors and monitor them.

Comment (nickheiner-usds) [reply app/controllers/annotation_controller.rb]: More broadly, why are we rescuing from `ActiveRecord::RecordInvalid` at all? Why not just have an uncaught exception that would be tracked automatically? Does the client actually use anything in the JSON we return here?

Comment (lowellrex) [reply app/controllers/annotation_controller.rb]: Looks like I made these particular 500s invisible due to [this PR](https://github.com/department-of-veterans-affairs/caseflow/pull/3856). In reality, these probably shouldn't be 500s at all, but rather 200s with an error element that we check on the front-end.

Comment (tejans24) [reply app/controllers/annotation_controller.rb]:
```
      (response) => {
        const responseObject = JSON.parse(response.response.text);

        dispatch({
          type: Constants.REQUEST_CREATE_ANNOTATION_FAILURE,
          payload: {
            annotationTemporaryId: temporaryId,
            errorMessage: responseObject.errors[0].detail
          }
        });
      }
```

Yes, it's being used. But I do think it makes sense to send a response if the request fails, i.e at-least return a 500 http status and let the client know something went wrong.

Comment (lowellrex) [reply app/controllers/annotation_controller.rb]: Do we have any guidelines on how we construct APIs?

Comment (nickheiner-usds) [reply app/controllers/annotation_controller.rb]: > In reality, these probably shouldn't be 500s at all, but rather 200s with an error element that we check on the front-end.

I think a `4xx` code could also be appropriate here, because trying to save an annotation with an empty content is a malformed request.

But why does the client make that request at all? Why do we let the user save an empty annotation if it's an error state?

Comment (nickheiner-usds) [reply app/controllers/annotation_controller.rb]: > Do we have any guidelines on how we construct APIs?

Not specifically, no.

Comment (nickheiner-usds) [reply app/controllers/annotation_controller.rb]: > Yes, it's being used but I'm not sure if the error message is relevant to the user.

Ok, thanks for sharing that.

We should be careful that we're not using a `500` here for a user input validation error. `500` should be reserved for actual exceptional situations.
